### PR TITLE
fix: support python 3.12

### LINKED ISSUE #9007: [CT-3325] [Tech Debt][Spike+] Test `dbt-core` with Python 3.12
### Housekeeping

- [X] I am a maintainer of dbt-core

### Short description

Python 3.12 was released last month, let's test that `dbt-core` works with this

https://www.python.org/downloads/release/python-3120/

### Acceptance criteria

* ~Bundles can be created with Python 3.12~
  * Removed since the current bundle integration tests are failing ([link](https://github.com/dbt-labs/dbt-core-bundles/actions/workflows/integration.yml)), will test with `dbt-core` 1.7.9 and 1.8.0b1
* dbt-common also supports python 3.12
* Update testing matrix to use python 3.12

### Impact to Other Teams

* upstream dependencies that will also need to be made compatible with 3.12: dbt-adapters dbt-semantic-interfaces
* 1p adapters

### Will backports be required?

No

### Context

https://docs.getdbt.com/faqs/Core/install-python-compatibility

“New dbt minor versions will add support for new Python3 minor versions as soon as all dependencies can support it. In turn, dbt minor versions will drop support for old Python3 minor versions right before they reach end of life.”

### PR DESCRIPTION
**Pull request by l1xnan**
resolves #9007

### Problem

`distutils.util.strtobool` was removed in Python 3.12, see https://peps.python.org/pep-0632

### Solution

Copying the `distutils.util.strtobool` implementation from python 3.10.

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me  
- [x] I have run this code in development and it appears to resolve the stated issue  
- [ ] This PR includes tests, or tests are not required/relevant for this PR
- [x] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc) or this PR has already received feedback and approval from Product or DX
- [x] This PR includes [type annotations](https://docs.python.org/3/library/typing.html) for new and modified functions

### CODE REVIEW

File: core/dbt/utils.py [hunk 1]
```diff
@@ -369,3 +369,19 @@ def args_to_dict(args):
...
+    are 'n', 'no', 'f', 'false', 'off', and '0'.  Raises ValueError if
+    'val' is anything else.
+    """
+    val = val.lower()
+    if val in ('y', 'yes', 't', 'true', 'on', '1'):
+        return 1
+    elif val in ('n', 'no', 'f', 'false', 'off', '0'):
+        return 0
+    else:
+        raise ValueError("invalid truth value %r" % (val,))
```
Comment (seb-posedio) [inline core/dbt/utils.py:387]: Missing newline

**Review by seb-posedio — CHANGES_REQUESTED** [review]

File: core/dbt/utils.py (hunk 1 above)
Comment (l1xnan) [reply core/dbt/utils.py]: Sorry, I've fixed it.

**Comment by aranke** [comment]
Hey @l1xnan,

Thanks for the PR, it looks good to me!

Can you please sign the CLA and add a changelog entry so we can merge this in?
Instructions on creating a changelog entry are here: https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#adding-or-modifying-a-changelog-entry

Thank you!

**Comment by l1xnan** [comment]
Run `pip install -e dbt-core` in python 3.11, it is installed by running `setup.py develop`:
```
# lib/site-packages
dbt/
dbt-core.egg-link
easy-install.pth    # The content is `<project-root>/dbt-core/core`
```

As a comparison, run `pip install -e dbt-core` in python 3.12, `dbt-core` is installed by building it as a temp wheel file:
```
# lib/site-packages
dbt/
dbt_core-1.8.0a1.dist-info/
__editable___dbt_core_1_8_0a1_finder.py
__editable__.dbt_core-1.8.0a1
```
The `easy-install.pth` file is missing, `dbt.include.starter_project` is only found in the `site-packages/dbt`, so the `dbt.include.starter_project` cannot be imported.

**Comment by l1xnan** [comment]
Official resolves

**Pull request closed without merging**